flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreReadImpl.java

```diff
@@ -20,60 +20,132 @@
 import org.apache.flink.table.store.file.mergetree.MergeTreeReader;
 import org.apache.flink.table.store.file.mergetree.compact.Accumulator;
 import org.apache.flink.table.store.file.mergetree.compact.IntervalPartition;
 import org.apache.flink.table.store.file.mergetree.sst.SstFileMeta;
 import org.apache.flink.table.store.file.mergetree.sst.SstFileReader;
 import org.apache.flink.table.store.file.utils.FileStorePathFactory;
 import org.apache.flink.table.store.file.utils.RecordReader;
+import org.apache.flink.table.store.utils.ProjectionUtils;
 import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nullable;
 
 import java.io.IOException;
+import java.util.Arrays;
 import java.util.Comparator;
 import java.util.List;
 
 /** Default implementation of {@link FileStoreRead}. */
 public class FileStoreReadImpl implements FileStoreRead {
 
+    private final RowType keyType;
     private final SstFileReader.Factory sstFileReaderFactory;
     private final Comparator<RowData> keyComparator;
     private final Accumulator accumulator;
 
+    private Projection<RowData, BinaryRowData> keyProjection;
+
     public FileStoreReadImpl(
             RowType keyType,
             RowType valueType,
             Comparator<RowData> keyComparator,
             Accumulator accumulator,
             FileFormat fileFormat,
             FileStorePathFactory pathFactory) {
+        this.keyType = keyType;
         this.sstFileReaderFactory =
                 new SstFileReader.Factory(keyType, valueType, fileFormat, pathFactory);
         this.keyComparator = keyComparator;
         this.accumulator = accumulator;
+
+        this.keyProjection = null;
     }
 
     @Override
     public void withKeyProjection(int[][] projectedFields) {
-        // TODO
-        throw new UnsupportedOperationException();
+        for (int[] projectedField : projectedFields) {
+            Preconditions.checkArgument(
+                    projectedField.length == 1,
+                    "FileStoreReadImpl currently does not support nested projection for keys");
+        }
+        keyProjection =
+                ProjectionUtils.newProjection(
+                        keyType, Arrays.stream(projectedFields).mapToInt(f -> f[0]).toArray());
     }
 
     @Override
     public void withValueProjection(int[][] projectedFields) {
-        // TODO
-        throw new UnsupportedOperationException();
+        sstFileReaderFactory.withValueProjection(projectedFields);
     }
 
     @Override
     public RecordReader createReader(BinaryRowData partition, int bucket, List<SstFileMeta> files)
             throws IOException {
-        return new MergeTreeReader(
-                new IntervalPartition(files, keyComparator).partition(),
-                true,
-                sstFileReaderFactory.create(partition, bucket),
-                keyComparator,
-                accumulator.copy());
+        RecordReader rawReader =
```

Comment:
If there is only one file, it is possible to support key projection directly.
We can have a separate branch.